src/java.base/share/classes/jdk/internal/lang/StableValue.java

```diff
@@ -0,0 +1,268 @@
+ *                 throw new IllegalArgumentException("Input not allowed: "+t);
+ *             }
+ *             return backing.get(t)
+ *                         .computeIfUnset(t, original);
+ *         };
+ *     }
+ * }
+ * <p>
+ * The constructs above are eligible for similar JVM optimizations as the StableValue
+ * class itself.
+ * <p>
+ * All methods that can set the stable value's holder value are guarded such that
+ * competing set operations (by other threads) will block if another set operation is
+ * already in progress.
+ * <p>
+ * Except for a StableValue's holder value itself, all method parameters must be
+ * <em>non-null</em> or a {@link NullPointerException} will be thrown.
+ *
+ * @param <T> type of the holder value
+ *
+ * @since 24
+ */
+public sealed interface StableValue<T>
+        permits StableValueImpl {
+
+    // Principal methods
+
+    /**
+     * {@return {@code true} if the holder value was set to the provided {@code value},
+     * otherwise returns {@code false}}
+     * <p>
+     * When this method returns, a holder value is always set.
+     *
+     * @param value to set (nullable)
+     */
+    boolean trySet(T value);
+
+    /**
+     * {@return the set holder value (nullable) if set, otherwise return the
+     * {@code other} value}
+     *
+     * @param other to return if the stable holder value is not set
+     */
+    T orElse(T other);
+
+    /**
+     * {@return the set holder value if set, otherwise throws {@code NoSuchElementException}}
+     *
+     * @throws NoSuchElementException if no value is set
+     */
+    T orElseThrow();
+
+    /**
+     * {@return {@code true} if a holder value is set, {@code false} otherwise}
+     */
+    boolean isSet();
+
+    /**
+     * If the holder value is unset, attempts to compute the holder value using the
+     * provided {@code supplier} and enters the result into the holder value.
+     *
+     * <p>If the {@code supplier} itself throws an (unchecked) exception, the exception
+     * is rethrown, and no holder value is set. The most common usage is to construct a
+     * new object serving as an initial value or memoized result, as in:
+     *
+     * {@snippet lang = java :
+     *     T t = stable.computeIfUnset(T::new);
+     * }
+     *
+     * @implSpec
+     * The implementation of this method is equivalent to the following steps for this
+     * {@code stable} and a given non-null {@code supplier}:
+     *
+     * {@snippet lang = java :
+     *     if (stable.isSet()) {
+     *         return stable.orElseThrow();
+     *     }
+     *     T newValue = supplier.get();
+     *     stable.trySet(newValue);
+     *     return newValue;
+     * }
+     * Except, the method is atomic, thread-safe and guarded with synchronization
+     * with respect to this method and all other methods that can set this StableValue's
+     * holder value.
+     * <p>
+     * If this method returns without throwing an Exception, a holder value is always set.
```

Comment:
I recommend moving this right above the `@implSpec` block tag; now it is considered part of that tag.